components/segmenter/src/provider/mod.rs

```diff
@@ -200,3 +201,68 @@ pub(crate) struct UCharDictionaryBreakDataV1Marker;
 impl DataMarker for UCharDictionaryBreakDataV1Marker {
     type Yokeable = UCharDictionaryBreakDataV1<'static>;
 }
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(databake::Bake),
+    databake(path = icu_segmenter::provider),
+)]
+/// Break state
+pub enum BreakState {
+    /// Break
+    Break,
+    /// Keep rule
+    Keep,
+    /// Non-matching rule
+    NoMatch,
+    /// We have to look ahead one more character.
+    Intermediate(u8),
+    /// Index of a property.
+    Index(u8),
+}
+
+#[cfg(feature = "datagen")]
+impl serde::Serialize for BreakState {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        // would be nice to use the derive serde for JSON, but can't break serialization
+        zerovec::ule::AsULE::to_unaligned(*self).serialize(serializer)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for BreakState {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        i8::deserialize(deserializer).map(zerovec::ule::AsULE::from_unaligned)
+    }
+}
+
+impl zerovec::ule::AsULE for BreakState {
```

Comment:
I like that this allows us to keep the bit-twiddling in one place.
Of course we will need to change it because we have more than 64 states in line breaking, and we need lookahead.

Could we get away with keeping it at 64 for sentence breaking and grapheme cluster breaking, and changing the representation only for line breaking?

Reply:
Oh, of course you are running into this already, which is why this breaks line break; you are deserializing some `Index` with high indices as `Intermediate` with different indices.

Reply:
Oh does line breaking *already* treat the bit patterns differently?

Reply:
Well, it does not have the intermediate break state bit-twiddling at all, which is how it can have more than 64 states.

You are adding that bit-twiddling in with this PR (though probably not the actual behaviour of looking ahead), which breaks things.

Reply:
So there are two different kinds of state interpretations...

Reply:
Indeed.

Reply:
> The unfortunate thing is just that the line segmenter uses the same values for extra states that the other segmenters use for intermediate states.

Only since #4389, so this has never been released.

The problem is that if we add add support for lookahead states to line breaking, the resulting state table, interpreted by earlier versions, will produce nonsense; whereas the state table from #4389 can be used with the pre-existing code and yields well-understood behaviour.

Reply:
Two issues there: 
1. anything resembling any standard or version of ICU beyond #4389 requires intermediate states;
2. those numbers are also indices into a table, so you would need to have a table with a big hole for 64 through 127 in order for that to work.